homeassistant/components/lyric/__init__.py

```diff
@@ -84,22 +88,34 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     )
 
     session = aiohttp_client.async_get_clientsession(hass)
-    oauth_session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
+    oauth_session = OAuth2SessionLyric(hass, entry, implementation)
 
     client = ConfigEntryLyricClient(session, oauth_session)
 
     client_id = hass.data[DOMAIN][CONF_CLIENT_ID]
     lyric = Lyric(client, client_id)
 
+    async def async_get_data() -> Lyric:
+        """Get data from the API."""
+        async with async_timeout.timeout(60):
+            await lyric.get_locations()
+        return lyric
+
     async def async_update_data() -> Lyric:
         """Fetch data from Lyric."""
-        await oauth_session.async_ensure_token_valid()
         try:
-            async with async_timeout.timeout(60):
-                await lyric.get_locations()
-            return lyric
-        except LyricAuthenticationException as exception:
-            raise ConfigEntryAuthFailed from exception
+            return await async_get_data()
+        except LyricAuthenticationException:
+            # Attempt to refresh the token before failing.
+            # Honeywell appear to have issues keeping tokens saved.
+            try:
+                _LOGGER.info("Authentication failed. Attempting to refresh token")
+                await oauth_session.force_refresh_token()
+                return await async_get_data()
+            except LyricAuthenticationException as exception:
+                raise ConfigEntryAuthFailed from exception
+            except (LyricException, ClientResponseError) as exception:
+                raise UpdateFailed(exception) from exception
```

Comment:
Instead of duplicating this all, you could call this method itself:

e.g.:

```py
    async def async_update_data(force_refresh_token: bool = False) -> Lyric:
        """Fetch data from Lyric."""
        try:
            if not force_refresh_token:
	            await oauth_session.async_ensure_token_valid()
	        else:
	            await oauth_session.force_refresh_token()
            return await async_get_data()
        except LyricAuthenticationException:
            if not force_refresh_token:
               return self.async_update_data(force_refresh_token=True)
            raise ConfigEntryAuthFailed from exception
        ...
```

Reply:
We may still want to catch `ClientResponseError` since our oauth2 helper may raise that. If the status code is 403 or 401 we should raise `ConfigEntryAuthFailed` otherwise `UpdateFailed`.